pkg/agent/openflow/network_policy.go

```diff
@@ -598,54 +704,159 @@ func (c *client) InstallPolicyRuleFlows(rule *types.PolicyRule) error {
+		action: &conjunctiveAction{
+			conjID:   c.id,
+			clauseID: clauseID,
+			nClause:  nClause,
+		},
+	}
+}
+
+// calculateClauses configures the policyRuleConjunction's clauses according to the PolicyRule. The Openflow entries are
+// not installed on the OVS bridge when calculating the clauses.
+func (c *policyRuleConjunction) calculateClauses(rule *types.PolicyRule, clnt *client) (uint8, binding.Table, binding.Table) {
+	var ruleTable, dropTable binding.Table
+	var isEgressRule = false
+	switch rule.Direction {
+	case v1.PolicyTypeEgress:
+		ruleTable = clnt.pipeline[egressRuleTable]
+		dropTable = clnt.pipeline[egressDefaultTable]
+		isEgressRule = true
+	default:
+		ruleTable = clnt.pipeline[ingressRuleTable]
+		dropTable = clnt.pipeline[ingressDefaultTable]
+	}
+
+	var fromID, toID, serviceID, nClause uint8
+	// Calculate clause IDs and the total number of clauses.
+	if rule.From != nil {
+		nClause += 1
+		fromID = nClause
+	}
+	if rule.To != nil {
+		nClause += 1
+		toID = nClause
+	}
+	if rule.Service != nil {
+		nClause += 1
+		serviceID = nClause
+	}
 
-	// Install conjunctive match flows if exists in rule.Form/To/Service
 	var defaultTable binding.Table
 	if rule.From != nil {
 		if isEgressRule {
 			defaultTable = dropTable
 		} else {
 			defaultTable = nil
 		}
-		conj.fromClause = conj.newClause(fromID, nClause, ruleTable, defaultTable)
-		if err := conj.fromClause.addAddrFlows(c, types.SrcAddress, rule.From); err != nil {
-			return err
-		}
+		c.fromClause = c.newClause(fromID, nClause, ruleTable, defaultTable)
 	}
 	if rule.To != nil {
 		if !isEgressRule {
 			defaultTable = dropTable
 		} else {
 			defaultTable = nil
 		}
-		conj.toClause = conj.newClause(toID, nClause, ruleTable, defaultTable)
-		if err := conj.toClause.addAddrFlows(c, types.DstAddress, rule.To); err != nil {
-			return err
-		}
+		c.toClause = c.newClause(toID, nClause, ruleTable, defaultTable)
 	}
 	if rule.Service != nil {
-		conj.serviceClause = conj.newClause(serviceID, nClause, ruleTable, nil)
-		if err := conj.serviceClause.addServiceFlows(c, rule.Service); err != nil {
-			return err
-		}
+		c.serviceClause = c.newClause(serviceID, nClause, ruleTable, nil)
 	}
-	c.policyCache.Store(rule.ID, conj)
-	return nil
+	return nClause, ruleTable, dropTable
 }
 
-func (c *policyRuleConjunction) newClause(clauseID uint8, nClause uint8, ruleTable, dropTable binding.Table) *clause {
-	return &clause{
-		ruleTable: ruleTable,
-		dropTable: dropTable,
-		matches:   make(map[string]*conjMatchFlowContext, 0),
-		action: &conjunctiveAction{
-			conjID:   c.id,
-			clauseID: clauseID,
-			nClause:  nClause,
-		},
+// getConjunctiveMatchFlowAddChange returns the conjMatchFlowContextChanges of the new policyRuleConjunction. It
+// will calculate the expected conjMatchFlowContex status, and the changed Openflow entries.
+func (c *policyRuleConjunction) getConjunctiveMatchFlowAddChange(clnt *client, rule *types.PolicyRule) []*conjMatchFlowContextChange {
```

Comment:
Do calculateChangesForRuleCreation and calculateChangesForRuleDeletion sound better?

Reply:
I will update.